[mlir][spirv] Add instruction OpGroupNonUniformRotateKHR

Hsiangkai opened the pull request:
Add an instruction under the extension SPV_KHR_subgroup_rotate.

The specification for the extension is here:
https://github.khronos.org/SPIRV-Registry/extensions/KHR/SPV_KHR_subgroup_rotate.html

Hsiangkai committed:
[mlir][spirv] Add instruction OpGroupNonUniformRotateKHR

Add an instruction under the extension SPV_KHR_subgroup_rotate.

The specification for the extension is here:
https://github.khronos.org/SPIRV-Registry/extensions/KHR/SPV_KHR_subgroup_rotate.html

kuhar commented:
@Hsiangkai can you merge this on your own or do you want me to do it?

IgWod-IMG commented:
It would be nice to have some verification or at least constrain operands and result types. For example:

> Result Type must be a scalar or vector of floating-point type, integer
> type, or Boolean type

```diff
- SPIRV_Type:$result
+ AnyTypeOf<[SPIRV_ScalarOrVectorOf<SPIRV_Float>, SPIRV_ScalarOrVectorOf<SPIRV_Integer>, SPIRV_ScalarOrVectorOf<SPIRV_Bool>]>:$result
```

(n.b. It would be nice to have `SPIRV_ScalarOrVectorOf` that takes a list of types but that's outside this PR)

Hsiangkai committed:
Add operand and result type constraint

Hsiangkai committed:
clang-format

Hsiangkai commented:
> It would be nice to have some verification or at least constrain operands and result types. For example:
> 
> > Result Type must be a scalar or vector of floating-point type, integer
> > type, or Boolean type
> 
> ```diff
> - SPIRV_Type:$result
> + AnyTypeOf<[SPIRV_ScalarOrVectorOf<SPIRV_Float>, SPIRV_ScalarOrVectorOf<SPIRV_Integer>, SPIRV_ScalarOrVectorOf<SPIRV_Bool>]>:$result
> ```
> 
> (n.b. It would be nice to have `SPIRV_ScalarOrVectorOf` that takes a list of types but that's outside this PR)

Thanks for your review.
I have added operand and result constraints and added operator verifier to check the validity of values. I also added several test cases to ensure the checking works.

kuhar commented on mlir/lib/Dialect/SPIRV/IR/GroupOps.cpp:
```diff
@@ -304,6 +304,38 @@ LogicalResult GroupNonUniformLogicalXorOp::verify() {
+  if (getDelta().getType().isSignedInteger())
+    return emitOpError("delta must be a singless/unsigned integer");
+
+  auto clusterSizeVal = getClusterSize();
```
Use the full type here since the type is not obvious based on the RHS.

Hsiangkai replied:
Done.

kuhar commented on mlir/lib/Dialect/SPIRV/IR/GroupOps.cpp:
```diff
@@ -304,6 +304,38 @@ LogicalResult GroupNonUniformLogicalXorOp::verify() {
+    return emitOpError("delta must be a singless/unsigned integer");
+
+  auto clusterSizeVal = getClusterSize();
+  if (clusterSizeVal) {
```
Since it's only used inside this if, you can do:
```suggestion
  if (TheType clusterSizeVal = getClusterSize()) {
```

IgWod-IMG commented on mlir/lib/Dialect/SPIRV/IR/GroupOps.cpp:
```diff
@@ -304,6 +304,36 @@ LogicalResult GroupNonUniformLogicalXorOp::verify() {
+  if (scope != spirv::Scope::Workgroup && scope != spirv::Scope::Subgroup)
+    return emitOpError("execution scope must be 'Workgroup' or 'Subgroup'");
+
+  if (getDelta().getType().isSignedInteger())
```
Could this check be avoided if `delta` is set to be  `SPIRV_SignlessOrUnsignedInt`? NB This is a good place to look at what types are already defined:

https://github.com/llvm/llvm-project/blob/e8711436b3419cc9e0e8a70c6eb41dbb2a1bf132/mlir/include/mlir/Dialect/SPIRV/IR/SPIRVBase.td#L4129

But I'm not going to lie, I sometimes personally get quite confused looking for the correct type :)

IgWod-IMG commented on mlir/lib/Dialect/SPIRV/IR/GroupOps.cpp:
```diff
@@ -304,6 +304,36 @@ LogicalResult GroupNonUniformLogicalXorOp::verify() {
+
+  auto clusterSizeVal = getClusterSize();
+  if (clusterSizeVal) {
+    if (clusterSizeVal.getType().isSignedInteger())
```
Same as above.

Hsiangkai replied:
Done.

IgWod-IMG commented on mlir/include/mlir/Dialect/SPIRV/IR/SPIRVNonUniformOps.td:
```diff
@@ -1361,4 +1361,78 @@ def SPIRV_GroupNonUniformBallotBitCountOp : SPIRV_Op<"GroupNonUniformBallotBitCo
+
+  let arguments = (ins
+    SPIRV_ScopeAttr:$execution_scope,
+    SPIRV_Type:$value,
```
Should we spell the type of the `$value` here to be `AnyTypeOf<[SPIRV_ScalarOrVectorOf<SPIRV_Float>, SPIRV_ScalarOrVectorOf<SPIRV_Integer>, SPIRV_ScalarOrVectorOf<SPIRV_Bool>]>`, which is the same type as a result?

I know the spec says nothing about the type directly and only defines it in terms of the result:

> The type of Value must be the same as Result Type.

But I had this discussion with @kuhar in the past: https://github.com/llvm/llvm-project/pull/124571#discussion_r1930755148 and it seems that we agreed that being more specific is better.

IgWod-IMG reviewed (commented):
I added some minor (optional) suggestion. Thank you for adding the verification!

Hsiangkai committed:
Address comments

IgWod-IMG commented on mlir/lib/Dialect/SPIRV/IR/GroupOps.cpp:
```diff
@@ -304,6 +304,29 @@ LogicalResult GroupNonUniformLogicalXorOp::verify() {
+  if (scope != spirv::Scope::Workgroup && scope != spirv::Scope::Subgroup)
+    return emitOpError("execution scope must be 'Workgroup' or 'Subgroup'");
+
+  if (TypedValue<Type> clusterSizeVal = getClusterSize()) {
```
It was suggested to me before to use `Value` instead of `TypedValue<Type>` in the situation like this: https://github.com/llvm/llvm-project/pull/126555#discussion_r1950950471 For what's worth I cannot see `TypedValue<Type>` being used anywhere in the dialect code.

Hsiangkai replied:
Thanks for your review. I have updated it. I will merge this patch after all tests pass.

IgWod-IMG reviewed (approved):
LGTM (with a small nitpick)

Hsiangkai committed:
Address comment
